src/modules/ecdsa_sign_to_contract/main_impl.h

```diff
@@ -0,0 +1,93 @@
+    const unsigned char* noncedata = NULL;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(signature != NULL);
+    ARG_CHECK(seckey != NULL);
+
+    if(s2c_data32 != NULL) {
+        /* Provide s2c_data32 to the the nonce function as additional data to derive the nonce
+         * from. It's hashed because it should be possible to derive nonces even if only a SHA256
+         * commitment to the data is known.  This is for example important in the
+         * anti-nonce-covert-channel protocol.
+         */
+        secp256k1_sha256 sha;
+        secp256k1_sha256_initialize(&sha);
+        secp256k1_sha256_write(&sha, s2c_data32, 32);
+        secp256k1_sha256_finalize(&sha, ndata);
+        noncedata = ndata;
+    }
+
+    ret = secp256k1_ecdsa_sign_inner(ctx, &r, &s, NULL, s2c_opening, msg32, seckey, s2c_data32, NULL, noncedata);
+    secp256k1_scalar_cmov(&r, &secp256k1_scalar_zero, !ret);
+    secp256k1_scalar_cmov(&s, &secp256k1_scalar_zero, !ret);
+    secp256k1_ecdsa_signature_save(signature, &r, &s);
+    return ret;
+}
+
+int secp256k1_ecdsa_s2c_verify_commit(const secp256k1_context* ctx, const secp256k1_ecdsa_signature *sig, const unsigned char *data32, const secp256k1_s2c_opening *opening) {
+    secp256k1_pubkey commitment;
+    secp256k1_ge commitment_ge;
+    unsigned char x_bytes1[32];
+    unsigned char x_bytes2[32];
+    secp256k1_scalar sigr, sigs;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(data32 != NULL);
+    ARG_CHECK(opening != NULL);
+    ARG_CHECK(secp256k1_s2c_commit_is_init(opening));
+
+    if (!secp256k1_ec_commit(ctx, &commitment, &opening->original_pubnonce, data32, 32)) {
+        return 0;
+    }
+
+    /* Check that sig_r == commitment_x (mod n)
+     * sig_r is the x coordinate of R represented by a scalar.
+     * commitment_x is the x coordinate of the commitment (field element).
+     *
+     * It is sufficient to only compare the x coordinates as it is as difficult to find an client
+     * commitment to the negation of the point as to any other point. There is a small reduction in
+     * security as it is easier to find a collision with a point and its negation.
+     */
+    secp256k1_ecdsa_signature_load(ctx, &sigr, &sigs, sig);
+
+    if (!secp256k1_pubkey_load(ctx, &commitment_ge, &commitment)) {
+        return 0;
+    }
+    secp256k1_fe_normalize(&commitment_ge.x);
+    secp256k1_fe_get_b32(x_bytes1, &commitment_ge.x);
+    secp256k1_scalar_get_b32(x_bytes2, &sigr);
+    return memcmp(x_bytes1, x_bytes2, 32) == 0;
+
+}
+int secp256k1_ecdsa_s2c_anti_nonce_covert_channel_host_commit(secp256k1_context *ctx, unsigned char *rand_commitment32, const unsigned char *rand32) {
+    secp256k1_sha256 sha;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(rand_commitment32 != NULL);
+    ARG_CHECK(rand32 != NULL);
+
+    secp256k1_sha256_initialize(&sha);
```

Comment:
tagged hash

Reply:
would need to be same tagged hash as `secdsa_s2c_sign` uses to create `noncedata` from the `s2c_data`.

Reply:
Do we want two separate tagged hashes, one for the data commitment and the other for the P→P+eG commitment?

Reply:
I think yes. We should separate domains as much as possible.